### DISCUSSION TITLE
How to train a custom model

### DISCUSSION THREAD

**Discussion by anhnh2002** [huggingface/peft#1559] [Q&A] [answered]
I must take some change llm architecture to adapt with my task.
How to train custom llm using peft?

**Comment by BenjaminBossan** [comment]
Custom models are generally supported, but you're not providing a lot of information (what adapter, what changes did you make?). Please take a look at our [docs on custom models](https://huggingface.co/docs/peft/developer_guides/custom_models) to see if those answer your question.

**Reply by anhnh2002** [reply]
I add some mlp layer to an LLM. And change training loop, sample data, loss func. Same in this repo https://github.com/nju-websoft/SCKD/tree/main .

**Reply by BenjaminBossan** [reply]
You can check the default `target_module` in [this file](https://github.com/huggingface/peft/blob/a18734d87aa9ae6b94b5bdde192b265bfad7c0b3/src/peft/utils/constants.py) and modify it according to your needs.

Let's use an example. Say your model is `gemma` and you want to use LoRA. You can see that the entry for `gemma` in `TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING` is `["q_proj", "v_proj"]`. Now let's say that your custom MLP module is called `my_mlp` (i.e that's its attribute name). If you want to apply LoRA to that layer as well, you should set `config = LoraConfig(..., target_modules=["q_proj", "v_proj", "my_mlp"])`. Alternatively, you can set `target_modules="all-linear"` to simply modify all `nn.Linear` layers.

In general, you can just print the model and see all the components and their names. Use this as a guidance for your choice of `target_modules`.

**Reply by anhnh2002** [reply]
Got it.
How can i change training loop, sampling data and loss func?

**Reply by BenjaminBossan** [reply, marked as answer]
PEFT doesn't prescribe any specific training code, use whatever you want. If you have some special things, it's probably easiest to write your own training loop with custom sampler and loss function.

**Reply by anhnh2002** [reply]
Sorry, my mistake

**Reply by anhnh2002** [reply]
Please @BenjaminBossan , i have a another question.
For example:
I have a llm with lora like this
```from transformers import AutoModelForSeq2SeqLM
from peft import get_peft_config, get_peft_model, LoraConfig, TaskType
model_name_or_path = "bigscience/mt0-large"
tokenizer_name_or_path = "bigscience/mt0-large"

peft_config = LoraConfig(
    task_type=TaskType.SEQ_2_SEQ_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.1
)

base_model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
model = get_peft_model(base_model, peft_config)
```

I can forward input by `model(input)`. This using base model's params and adapter's params.
How can i forward input without using adapter's params? I want to forward input with and without adapter's params with only one base model loading

**Reply by BenjaminBossan** [reply]
Use the `disable_adapter` context like this:

```python
with model.disable_adapter():
    output = model(input)
```

**Reply by anhnh2002** [reply]
thank you so much

**Reply by anhnh2002** [reply]
Please @BenjaminBossan, another question.
For example:

- I have base model M and two adapter weights A and B.

With only one base model loading. How to forward input like this: `(M+A)(input)` and `(M+B)(input)` where A consist trainable params and B consist untrainable params?

**Reply by BenjaminBossan** [reply]
So I assume here that you mean two different adapters, not something like the `lora_A` weight and the `lora_B` weight of one adapter.

After loading the `PeftModel` with `get_peft_model`, you can add a 2nd adapter using either `model.load_adapter(...)` for already trained adapters, or `model.add_adapter(...)` to add a new, untrained adapter. To switch between the two, use `model.set_adapter(<name>)`.

To make adapters untrainable, set their `requires_grad` attribute to `False`, like you would do with PyTorch in general.

**Reply by anhnh2002** [reply]
Thank a lot
